docs/proposals/20190709-cluster-spec-crds.md

```diff
@@ -0,0 +1,325 @@
+- Embedding the provider status as part of the Cluster requires that either the generic cluster controller be responsible for pulling this state from the provider (e.g. by invoking a provider actuator) or the provider updating this field, incurring in overlapping responsibilities. 
+
+### Goals
+
+1. To use Kubernetes Controllers to manage the lifecycle of provider specific cluster infrastructure object.
+1. To validate provider-specific object as early as possible (create, update).
+1. Allow cluster providers to expose the status via a status subresource, eliminating the need of status updates across controllers.
+
+### Non-Goals/Future Work
+1. To modify the generic Cluster object beyond the elements related to the provider-specific specs and status.
+1. Splitting the provider-specific infrastructure into more granular elements, such as network infrastructure and control plane specs. Revisiting the definition of the provider specific infrastructure is left for further proposals.
+1. To replace the cluster status by alternative representations, such as a list of conditions.
+1. To propose cluster state lifecycle hooks. This must be part of a future proposal that builds on these proposed changes
+
+## Proposal
+
+This proposal introduces changes in the Cluster data model and describes a collaboration model between the Cluster controller and the provider infrastructure controller, including a model for managing the state transition during the provisioning process. 
+
+### Data Model changes
+
+```go
+type ProviderSpec struct
+```
+- **To remove**
+    - **Value** [optional] _Superseded by InfrastructureRef_
+        - Type: `*runtime.RawExtension`
+        - Description: Value is an inlined, serialized representation of the resource configuration. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field, akin to component config.
+
+- **To add**
+    - **InfrastructureRef** [optional]
+        - Type: `*corev1.ObjectReference`
+        - Description: InfrastructureRef is a reference to a provider-specific resource that holds the details for provisioning a cluster in said provider. This reference is optional and allows the provider to implement its own cluster specification details.
+
+```go
+type ClusterStatus struct
+```
+- **To remove**
+    - **ProviderStatus** [optional]
+        - Type: `*runtime.RawExtension`
+        - Description:  Provider-specific status. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field.
+
+- **To add**
+    - **InfrastructureStatusRef** [optional]
```

Comment:
I don't think this is needed, the Reference in Spec should be sufficient to get the status as well